### ISSUE TITLE
Fix date parsing regression introduced in 2018

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#6009]
The current behavior of interpreting date strings without a timezone as being in UTC was a breaking change [introduced](https://github.com/OpenRefine/OpenRefine/commit/19f98b7ea2a012065a9e21999a9f9dc30347daf7#diff-665d024fa717cd5b5954821675bb5fafae) in May 2018 without any announcement (as far as I can tell). The fact the [tests were changed](https://github.com/OpenRefine/OpenRefine/commit/19f98b7ea2a012065a9e21999a9f9dc30347daf7#diff-e5e66f9d10edc76fb1e0e9fe53fb4770ab7f8a1e6ea4f76b7e5d36ad56dd3d4eR31-R32) should have been a red flag that something bad was going on.

I first discovered this in 2020 and added a [ToDo](https://github.com/OpenRefine/OpenRefine/commit/55edae2b7bb835bd2ed87cd8ab35e49d64c41767#diff-55584f092025686d30ae6ff51c8e7f9e6d7e42ddba2c0e0341b8345368fcdf65R178-R180) about it to the relevant tests and mentioned it in the comments to #3027, but apparently didn't create an issue, so remedying that now.

### Current Results
All dates and datetime strings are interpreted as being at UTC when parsed.

### Expected Behavior
I believe we should go back to the historical and, to me, more logical behavior of interpreting dates as being in the local timezone rather than as being at UTC.

### Versions
OpenRefine <=2.7 - correct 
OpenRefine 3.0+ - broken

### Additional context
This is a second breaking change to undo the first breaking change, which is less than ideal, but I believe it's also more logical to the user to have things interpreted in the local timezone.

**Comment by wetneb** [comment]
I agree this change was not very helpful, I feel like it was mostly motivated by an interest in upgrading to newer Java APIs without much consideration of the user's requirements.

As a user, I see more options than "interpreting in local time" or "interpreting in UTC". There is a third option, which is "interpreting as a date(time) without clear timezone info". I think that third option should be the default behaviour for a data cleaning tool, especially for dates (where timezones are even more rarely specified).

This phrasing of "interpreting" might also be a bit abstract and hard to relate to concrete user workflows, so it's worth trying to pin it down in concrete terms. For instance, when loading a CSV file with datatype auto-detection:
```
person,birthdate
Maribel,1985-07-04
Fernando,1976-12-22
Jose,1988-03-08
```
I want that the year, month and day components of the dates to be exactly what they were in the source file, and I want them to stay the same afterwards, even if I created the project while being in the UTC+2 timezone and re-opened it later while being in the UTC-5 timezone. I want those dates to be preserved in OpenRefine and also if I export the project to any format.

I would expect the same behaviour if parsing datetime objects with no timezone attached (such as `1989-09-06 08:34:45`).

In my opinion, this is different from interpreting date(time) objects in the local timezone, because if we do so, then the rendering of those objects should change depending of that local timezone (to preserve the interpretation).

**Comment by ostephens** [comment]
I made this comment on the forum as well but repeating here:

>The advantages I can see to the current behaviour is that it offers reproducibility and consistency between users - which is useful in training, support and in cases where users are working collaboratively but are geographically across time zones (or a single user working on the same data inside/outside daylight saving time? I'm not sure how the previous behaviour handled this situation?).

> Outside this I don't have strong feelings on the issue - the most important thing is that we are clear about the behaviour

> I do like the suggestion that we have more/better tools for converting date times between time zones - possibly on import, on export (as suggested by [@Tim_Chaffe](https://forum.openrefine.org/u/tim_chaffe)) and in transformations.

**Comment by tfmorris** [comment]
OK, I clearly should have used a better title. Hopefully the updated title is clearer. This is only about fixing the regression, not adding new functionality. The only question in my mind is whether fixing the regression is worse than leaving it as is.

> As a user, I see more options than "interpreting in local time" or "interpreting in UTC". There is a third option, which is "interpreting as a date(time) without clear timezone info". I

Sure, but that's a significant chunk of new functionality to implement, and making it the default would not be backwards compatible. We have #533, which I reopened last year, to cover fixing cases where extra information is being unnecessarily added/derived.

> I agree this change was not very helpful, I feel like it was mostly motivated by an interest in upgrading to newer Java APIs without much consideration of the user's requirements.

But changing to the new APIs added no technical benefit **OR** user benefit and, additionally, was done incorrectly making the new code more confusing because it misuses the APIs. This was all done by a technically incompetent employee of a commercial company that granted them access to the source repository because they had admin rights. If I could find a way, I'd charge that company for every hour I spend repairing this mess. It's an infuriating waste of time.

**Comment by wetneb** [comment]
Tom, I understand your frustration but I think there is no point in escalating the blame on this contributor who has long left the project by now. A lot of his contributions were made in a personal capacity. Let's stay focused on the possible solutions to the problem and keep the discussion constructive.

**Comment by tfmorris** [comment]
Any opinions on my question?

> The only question in my mind is whether fixing the regression is worse than leaving it as is.

**Comment by wetneb** [comment]
I don't know exactly what you mean by "fixing the regression". In your first message you propose to "interpret in local time" but I am not sure I understand what you mean exactly by that, which is why I tried to phrase the problem differently in my first reply above. I think there are multiple approaches, they could each be better or worse than the status quo:
* attempting a plain revert, i.e. something as close as possible to a `git revert` modulo the more recent changes, which would amount going back to the `Calendar` class internally
* migrating to something else from the Java 8+ API, such as `LocalDateTime`, which embeds a timezone I think, so it's perhaps not a great fit
* some home-grown date(time) class made to fit our needs
* something else?

In any case the move should be motivated by concrete use cases where we'd be able to validate the improvement from a user perspective.

**Comment by thadguidry** [comment]
I would fix regression.  Users are currently not aware of what is happening most of the time with their datetimes.  We should make them aware then give them the opportunity at import to make a choice.  Maybe a dialog? The smart default should be local time and we should inform the user of this at import. I have never supported keeping hidden any transformations OpenRefine does, and that's the current situation that should be fixed.  So can OpenRefine inform the user at import?

**Comment by ostephens** [comment]
> The only question in my mind is whether fixing the regression is worse than leaving it as is.

As you say @tfmorris "damned if you do, damned if you don't" . Given this my vote is:

* leave the existing functionality as it is
* make sure this is clearly documented (suggest this includes a note of the former behaviour and which version this changed in)
* put effort into improving our date/time handling including timezone options

**Comment by tfmorris** [comment]
@ostephens @thadguidry I'm counting on you to follow through with issues/documentation that implement your proposed solution. I've dropped my work on a code based solution.

**Comment by thadguidry** [comment]
@tfmorris @ostephens Please review my new PR which adds an admonition to our docs in both GREL Reference for Dates and the Dates user guide.  Hopefully I've captured the provenance well enough, and noted one benefit for users.
